### ISSUE TITLE
Serenity/JS 3.0 - ERROR webdriver: Request failed with status 404 due to no such alert

### ISSUE THREAD

**Issue by viper3400** [serenity-js/serenity-js#1156]
When using `webdriver` instead of `devtools` with the WDIO integration the same error is logged multiple times. It seams at each interaction which is performed via webdriver.

```
ERROR webdriver: Request failed with status 404 due to no such alert
```

https://developer.mozilla.org/en-US/docs/Web/WebDriver/Errors#table_of_errors

Test will run and does not fail. Do you know, what this error stands for and why it is raised? It’s a bit disturbing. I dit not test, yet, if this error would appear in a pure WDIO setup as well (without using Serenity/JS).

*Reproduction*
- Switch to the Serenity/JS main branch
- Run `npm ci`, `npm run lerna:bootstrap:all`, `npm run compile`
- `cd` into `./integration/webdriverio-web`
- In the `wdio.conf` change `automationProtocol` to `webdriver`
- In a terminal run `npm run start`, in an other terminal run `npx  chromedriver --port=4444`
- In another terminal run `npm run test:acceptance:run`

**Comment by viper3400** [comment]
@jan-molak, can you have a look at this? I was not able to reproduce this with pure WDIO + webdriver protocol. Also not sure if this was already an issue with WDIO integration in 2.x or if this is new to 3.x.

**Comment by viper3400** [comment]
@jan-molak , any news here? I tested this with https://github.com/serenity-js/serenity-js/tree/master/examples/webdriverio-mocha-todomvc and can't reproduce it there. So I assume, it's an issue with the 3.x integration,

**Comment by jan-molak** [comment]
Hey @viper3400, so the error shows up when running `ModalDialog.spec.ts`, which you can see by changing the config in `integration/webdriverio-web/wdio.conf.ts` as follows:

```typescript
const specs = [
    // './node_modules/@integration/web-specs/spec/**/*.spec.ts',
    
    // Error logged
    './node_modules/@integration/web-specs/spec/**/ModalDialog.spec.ts',

    // Error not logged    
    './node_modules/@integration/web-specs/spec/**/PageElement.located.ByCss.spec.ts',
];
```

I think the error is caused by the this logic in `packages/webdriverio/src/screenplay/models/WebdriverIOModalDialog.ts`:

```typescript
    isPresent(): Promise<boolean> {
        return this.browser.getAlertText()
            .then(() => true, error => {
                if (error && (
                    error.name === 'no such alert'
                    || error.message.startsWith('no such alert')
                )) {
                    return false;
                }

                throw error;
            });
    }
```

To my knowledge, the only way to check if an alert is present using the WebDriver protocol is to try to switch to it and check if that operation throws an error. If it threw an error called `no such alert`, the alert is not present, if it didn't throw an error, the alert is present.

While this is handled on the Serenity/JS side, WebDriver server itself still throws an error on its side, and that's what WebdriverIO prints to the console.

I don't think there's a way around it unless I missed a more elegant way to do it 🤔

**Comment by viper3400** [comment]
Hi @jan-molak, thanks for having a look into that. I'm pretty sure that we do not use `ModalDialog` in the project where I've seen this error first (and there are a lot of it in the logs). Obviously I was wrong about reproducibility steps.

Next try:

- Clone the https://github.com/serenity-js/serenity-js-mocha-webdriverio-template and switch to the `main` branch.
- Change `automationProtcol` in `wdio.conf` to `webdriver`
- `npx chromedriver --port=4444` (assuming you've one installed)
- `npm run test`

I've attached the [output](https://github.com/serenity-js/serenity-js/files/8500126/1156-console-out.txt)

There are just three steps in the `example.spec.ts` but it seems each of them throws this error. There is no `ModalDialog` and even the `Click.on` a `PageElement` seems to throw this error.

Btw, what about this retry thing. I can raise the `connectionRetryCount` in the `wdio.conf` to `100` and it seems it would iterate 100 times than (before the error is thrown)

```
[test:execute] [0-0] 2022-04-16T15:00:27.857Z INFO webdriver: Retrying 1/100
[test:execute] [0-0] 2022-04-16T15:00:27.857Z INFO webdriver: [GET] http://localhost:4444/session/5be59ea06c018ae8da261378921bb0b5/alert/text

[....]

[test:execute] [0-0] 2022-04-16T15:00:27.857Z INFO webdriver: Retrying 99/100
[test:execute] [0-0] 2022-04-16T15:00:27.857Z INFO webdriver: [GET] http://localhost:4444/session/5be59ea06c018ae8da261378921bb0b5/alert/text
[test:execute] [0-0] 2022-04-16T15:00:27.859Z DEBUG webdriver: request failed due to response error: no such alert
[test:execute] [0-0] 2022-04-16T15:00:27.859Z WARN webdriver: Request failed with status 404 due to no such alert
[test:execute] [0-0]   (Session info: chrome=100.0.4896.127)
[test:execute] [0-0] 2022-04-16T15:00:27.859Z INFO webdriver: Retrying 100/100
[test:execute] [0-0] 2022-04-16T15:00:27.859Z INFO webdriver: [GET] http://localhost:4444/session/5be59ea06c018ae8da261378921bb0b5/alert/text
[test:execute] [0-0] 2022-04-16T15:00:27.862Z DEBUG webdriver: request failed due to response error: no such alert
[test:execute] [0-0] 2022-04-16T15:00:27.862Z ERROR webdriver: Request failed with status 404 due to no such alert: no such alert
```

**Comment by viper3400** [comment]
Hi @jan-molak, today I removed the `Photographer.whoWill(TakePhotosOfInteractions)` from my `crew` array int the `wdio.conf` for some other reason.

But then it happens, that this error message was gone. Looks like there is a dependency to the screenshots!?

**Comment by jan-molak** [comment]
Hey @viper3400! The `Photographer` won't be able to take the screenshot if there's a modal dialog open, which is a limitation of the webdriver protocol. So before taking the screenshot, it [checks if there is a modal dialog open](https://github.com/serenity-js/serenity-js/blob/8bd28e99ef21ce0cf44e9c95e64761b0a89196fa/packages/web/src/stage/crew/photographer/strategies/PhotoTakingStrategy.ts#L53-L64), and if there is - it aborts taking the screenshot.

Annoyingly, the only way to check if the modal is open is to switch it, and trigger an error if there's no modal.

An alternative approach would be to attempt taking a screenshot, and should that operation fail - check if the error was caused by a modal being open. Unfortunately, each web integration tool (Protractor, WebdriverIO, Playwrightm etc.) uses a slightly different error message to communicate that, so we'd have to leak this knowledge to [`PhotoTakingStrategy`](https://github.com/serenity-js/serenity-js/blob/8bd28e99ef21ce0cf44e9c95e64761b0a89196fa/packages/web/src/stage/crew/photographer/strategies/PhotoTakingStrategy.ts#L107-L111). Or, we could have implementation-specific Photographers...

**Comment by jan-molak** [comment]
Hey @viper3400!

`3.0.0-rc.20` introduces a new model of working with alerts, which should also address the issue you were seeing. Please have a look at [`ModalDialog.spec.ts`](https://github.com/serenity-js/serenity-js/blob/main/integration/web-specs/spec/screenplay/models/ModalDialog.spec.ts) and let me know what you think.

In short, we tell the actor how they should react to the next dialog they see (so basically we set a handler), and then assert on the state of the dialog post interaction:

```typescript
            actorCalled('Nick').attemptsTo(
                ModalDialog.acceptNextWithValue('my name is Nick'),

                Click.on(Example.trigger),

                Ensure.that(ModalDialog.lastDialogState(), equals('accepted')),
            )
```

This way we have a consistent API for Playwright, Puppeteer, WebdriverIO, and Selenium.

**Comment by jan-molak** [comment]
🚀 New implementation of `ModalDialog` is available as of 3.0.0-rc.20. It doesn't rely on checking if the modal is present before switching windows, so won't be causing the WebDriver error.

**Comment by viper3400** [comment]
Thanks for fixing this, it works.
